src/mds/Server.cc

```diff
@@ -5427,11 +5427,95 @@ void Server::handle_client_setdirlayout(MDRequestRef& mdr)
 }
 
 // XATTRS
+std::pair<int, Server::JSONParsingError> Server::parse_layout_vxattr_json(
+  string name, string value, const OSDMap& osdmap, file_layout_t *layout)
+{
+  auto parse_pool = [&](std::string pool_name, int64_t pool_id) -> int64_t {
+    if (pool_name != "") {
+      int64_t _pool_id = osdmap.lookup_pg_pool_name(pool_name);
+      if (_pool_id < 0) {
+	dout(10) << __func__ << " unknown pool name:" << pool_name << dendl;
+	return -CEPHFS_EINVAL;
+      }
+      return _pool_id;
+    } else if (pool_id >= 0) {
+      const auto pools = osdmap.get_pools();
+      if (pools.find(pool_id) == pools.end()) {
+	dout(10) << __func__ << " unknown pool id:" << pool_id << dendl;
+	return -CEPHFS_EINVAL;
+      }
+      return pool_id;
+    } else {
+      return -CEPHFS_EINVAL;
+    }
+  };
 
-int Server::parse_layout_vxattr(string name, string value, const OSDMap& osdmap,
-				file_layout_t *layout, bool validate)
+  try {
+    if (name == "layout") {
+      JSONParser json_parser;
+      if (json_parser.parse(value.c_str(), value.length()) and json_parser.is_object()) {
+	std::string field;
+	try {
+	  field = "object_size";
+	  JSONDecoder::decode_json("object_size", layout->object_size, &json_parser, true);
+
+	  field = "stripe_unit";
+	  JSONDecoder::decode_json("stripe_unit", layout->stripe_unit, &json_parser, true);
+	  
+	  field = "stripe_count";
+	  JSONDecoder::decode_json("stripe_count", layout->stripe_count, &json_parser, true);
+	  
+	  field = "pool_namespace";
+	  JSONDecoder::decode_json("pool_namespace", layout->pool_ns, &json_parser, false);
+	  
+	  field = "pool_id";
+	  int64_t pool_id = 0;
+	  JSONDecoder::decode_json("pool_id", pool_id, &json_parser, false);
+
+	  field = "pool_name";
+	  std::string pool_name;
+	  JSONDecoder::decode_json("pool_name", pool_name, &json_parser, false);
+
+	  pool_id = parse_pool(pool_name, pool_id);
+	  if (pool_id < 0) {
+	    return {(int)pool_id, JSONParsingError::BAD_POOL_ID};
+	  }
+	  layout->pool_id = pool_id;
+	} catch (JSONDecoder::err&) {
+	  dout(10) << __func__ << " json is missing a mandatory field named " << field << dendl;
+	  return {-CEPHFS_EINVAL, JSONParsingError::MISSING_MANDATORY_FIELD};
+	}
+      } else {
+	dout(10) << " bad json" << dendl;
+	return {-CEPHFS_EINVAL, JSONParsingError::BAD_JSON};
+      }
+    } else if (name == "layout.object_size") {
+      layout->object_size = boost::lexical_cast<unsigned>(value);
+    } else if (name == "layout.stripe_unit") {
+      layout->stripe_unit = boost::lexical_cast<unsigned>(value);
+    } else if (name == "layout.stripe_count") {
+      layout->stripe_count = boost::lexical_cast<unsigned>(value);
+    } else if (name == "layout.pool_id") {
+      layout->pool_id = parse_pool(""s, boost::lexical_cast<int64_t>(value));
+    } else if (name == "layout.pool" or name == "layout.pool_name") {
```

Comment:
use `||` (doing too much python? ;)

Reply:
the [keyword `or`](https://en.cppreference.com/w/cpp/keyword/or) is a valid C++ token representing the logical OR operator which can be used in lieu of the `||` token
we are already at C++17 ... and shouldn't be shying away from such features
any specific reason against the use of `or` ?

Update:
The Ceph CodingStyle document points to the Google C++ Coding Style document which insists on using `&&` instead of `and`, but nothing about `or`. I think that should be taken as a generic statement about using punctuation operators (eg. `||`) instead of word operators (eg. `or`)

I've changed the word operator `or` to punctuation operator `||`

Reply:
Right - but I've not seen this being extensively followed in ceph codebase::

```
git grep " or " | grep "\.cc"
```

All are comments in source files.